electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,266 @@
+from typing import List
+import logging
+
+logging.basicConfig(level=logging.INFO)
+
+import random
+import os
+from tqdm import tqdm
+import json
+import argparse
+import numpy as np
+
+from schrodinger.structure import StructureReader
+from schrodinger.structutils import analyze
+from solvation_shell_utils import (
+    expand_shell,
+    renumber_molecules_to_match,
+    filter_by_rmsd,
+    filter_shells_with_solute_atoms,
+)
+from utils import validate_metadata_file
+from schrodinger.comparison import are_conformers
+from schrodinger.application.jaguar.utils import group_with_comparison
+from schrodinger.structutils import rmsd
+
+
+def extract_solvation_shells(
+    input_dir: str,
+    save_dir: str,
+    system_name: str,
+    solute_radii: List[float],
+    solvent_radii: List[float],
+    top_n: int,
+):
+    """
+    Given a MD trajectory in a PDB file, perform a solvation analysis
+    on the specified solute to extract the first solvation shell.
+
+    Args:
+        input_dir: Path to 1) the PDB file containing the MD trajectory (system_output.pdb) and 2) a metadata file (system_metadata.json)
+        save_dir: Directory in which to save extracted solvation shells.
+        system_name: Name of the system - used for naming the save directory.
+        solute_radii: List of shell radii to extract around solutes.
+        solvent_radii: List of shell radii to extract around solvents.
+        top_n: Number of snapshots to extract per topology.
+    """
+
+    # Read a structure and metadata file
+    # TODO: add charges
+    logging.info("Reading structure and metadata files")
+
+    # Read metadata
+    with open(os.path.join(input_dir, "metadata_system.json")) as f:
+        metadata = json.load(f)
+
+    validate_metadata_file(metadata)
+
+    partial_charges = np.array(metadata["partial_charges"])
+
+    solutes = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solute"
+    }  # {species: residue name} mapping
+    solute_resnames = list(solutes.values())
+
+    solvents = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solvent"
+    }  # {species: residue name} mapping
```

Comment:
In ionic liquid systems, is everything being treated as a "solute", or are the "main" salts being considered as "solvents"?

Reply:
I believe everything is being treated as a solute. In that case, the current way `expand_shells` is written might not work, because it'll always keep expanding until the size limit. Let's discuss this tomorrow.

Reply:
I'll make a note to bring this up.

Reply:
Marking as resolved: we decided to not do expansion at all for solvent-centered shells.